Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1

```diff
@@ -414,23 +422,77 @@ function Wait-ServiceFabricApplicationTypeTerminalStatus
         -RetryMessage (Get-VstsLocString -Key SFSDK_RetryingGetApplicationType)
 }
 
-function Unregister-ServiceFabricApplicationTypeAction
+function Wait-ServiceFabricApplicationTypeRegistrationStatus
 {
-    Param(
+    Param (
         [string]
         $ApplicationTypeName,
 
         [string]
         $ApplicationTypeVersion,
 
         [int]
-        $TimeoutSec
+        $TimeoutSec,
+
+        [string]
+        $OperationType
+    )
+
+    $global:operationId = $SF_Operations.GetApplicationType
+    $getAppTypeAction = { Get-ServiceFabricApplicationType -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion }
+    $getAppTypeRetryEvaluator = {
+        param($appType)
+        # if app type does not exist (i.e it got unprovisioned) or if its status has changed to a terminal one, stop wait
+        if ((!$appType) -or (($appType.Status -ne [System.Fabric.Query.ApplicationTypeStatus]::Provisioning) -and ($appType.Status -ne [System.Fabric.Query.ApplicationTypeStatus]::Unprovisioning)))
+        {
+            return $false
+        }
+        else
+        {
+            Write-Host (Get-VstsLocString -Key SFSDK_ApplicationTypeStatus -ArgumentList @($appType.Status, $appType.StatusDetails))
+            return $true
+        }
+    }
+
+    $MaxTries = 1200
+    $RetryIntervalInSeconds = 3
+    if($TimeoutSec)
+    {
+        $MaxTries = [int]($TimeoutSec/$RetryIntervalInSeconds)
+    }
+
+    $timeoutAction = $null
+    if($OperationType -eq $SF_Operations.RegisterApplicationType)
+    {
+        $timeoutAction = {
```

Comment:
Why add an action for this?
You can just put Invoke-ActionWithRetries under try-catch and perform these steps on TimeoutException

Reply:
Yes we can but this is more clear, readable and the word TIMEOUTACTION make more sense to categorize the action that is being performed.
Also this might be helpful for future as well